lib/libc/gen/getttyent.3
.\"	BSDI getttyent.3,v 2.5 1997/01/15 20:02:53 bostic Exp
.\"
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)getttyent.3	8.1 (Berkeley) 6/4/93
.\"
.Dd Aug 1, 1995
.Dt GETTTYENT 3
.Os BSD 4.3
.Sh NAME
.Nm getttyent ,
.Nm getttynam ,
.Nm setttyent ,
.Nm endttyent
.Nd get ttys file entry
.Sh SYNOPSIS
.Fd #include <ttyent.h>
.Ft struct ttyent *
.Fn getttyent void
.Ft struct ttyent *
.Fn getttynam "const char *name"
.Ft int
.Fn setttyent void
.Ft int
.Fn endttyent void
.Sh DESCRIPTION
The
.Fn getttyent ,
and
.Fn getttynam
functions
each return a pointer to an object, with the following structure,
containing the broken-out fields of a line from the tty description
file.
.Bd -literal
struct ttyent {
	char	*ty_name;	/* terminal device name */
	char	*ty_getty;	/* command to execute */
	char	*ty_type;	/* terminal type */
	int	ty_status;	/* flag values */
	char	*ty_window;	/* command for window manager */
	char	*ty_comment;	/* comment field */
	char	*ty_auth;	/* authentication mode */
};
.Ed
.Pp
The fields are as follows:
.Bl -tag -width ty_comment
.It Fa ty_name
The name of the character-special file.
.It Fa ty_getty
The name of the command invoked by
.Xr init 8
to initialize tty line characteristics.
.It Fa ty_type
The name of the default terminal type connected to this tty line.
.It Fa ty_status
A mask of bit fields which indicate various actions allowed on this
tty line.
The possible flags are as follows:
.Bl -tag -width TTY_SECURE
.It Dv TTY_ON
Enables logins (i.e.,
.Xr init 8
will start the command referenced by
.Fa ty_getty
on this entry).
.It Dv TTY_SECURE
Allow users with a uid of 0 to login on this terminal.
.\" .It Dv TTY_DIALOUT
.\" Dialout modem line controlled by dialer daemon.
.\" .It Dv TTY_DIALIN
.\" Dialin modem line controlled by dialer daemon.
.\" .It Dv TTY_BIDIR
.\" Bidirectional modem line controlled by dialer daemon.
.El
.It Fa ty_window
The command to execute for a window system associated with the line.
.It Fa ty_comment
Any trailing comment field, with any leading hash marks (``#'') or
whitespace removed.
.It Fa ty_auth
The authentication type to be used for this line.
See
.Xr login.con 5 .
.El
.Pp
If any of the fields pointing to character strings are unspecified,
they are returned as null pointers.
The field
.Fa ty_status
will be zero if no flag values are specified.
.Pp
See
.Xr ttys 5
for a more complete discussion of the meaning and usage of the
fields.
.Pp
The
.Fn getttyent
function
reads the next line from the ttys file, opening the file if necessary.
The
.Fn setttyent
function
rewinds the file if open, or opens the file if it is unopened.
The
.Fn endttyent
function
closes any open files.
.Pp
The
.Fn getttynam
function
searches from the beginning of the file until a matching
.Fa name
is found
(or until
.Dv EOF
is encountered).
.Sh RETURN VALUES
The routines
.Fn getttyent
and
.Fn getttynam
return a null pointer on
.Dv EOF
or error.
The
.Fn setttyent
function
and
.Fn endttyent
return 0 on failure and 1 on success.
.Sh FILES
.Bl -tag -width /etc/ttys -compact
.It Pa /etc/ttys
.El
.Sh SEE ALSO
.Xr ttyslot 3 ,
.Xr gettytab 5 ,
.Xr login.conf 5 ,
.Xr termcap 5 ,
.Xr ttys 5 ,
.\" .Xr dialer 8
.Xr getty 8 ,
.Xr init 8 ,
.Xr login 8
.Sh HISTORY
The
.Fn getttyent ,
.Fn getttynam ,
.Fn setttyent ,
and
.Fn endttyent
functions appeared in
.Bx 4.3 .
.Sh BUGS
These functions use static data storage;
if the data is needed for future use, it should be
copied before any subsequent calls overwrite it.
.Pp
The
.Fn getttyent ,
.Fn getttynam ,
.Fn setttyent ,
and
.Fn endttyent
functions may not be safely called concurrently from multiple 
threads, e.g., the interfaces described by 
.Fn pthreads 3 .


lib/libc/gen/getttyent.c
/*	BSDI getttyent.c,v 2.5 2000/04/03 16:21:17 donn Exp	*/

/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)getttyent.c	8.1 (Berkeley) 6/4/93";
#endif /* LIBC_SCCS and not lint */

#include <ctype.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <ttyent.h>

static char zapchar;
static FILE *tf;

static char *skip __P((char *));
static char *value __P((char *));

struct ttyent *
getttynam(tty)
	const char *tty;
{
	register struct ttyent *t;

	setttyent();
	while ((t = getttyent()) != NULL)
		if (!strcmp(tty, t->ty_name))
			break;
	endttyent();
	return (t);
}

struct ttyent *
getttyent()
{
	static struct ttyent tty;
	register int c;
	register char *p;
	size_t len;

	if (!tf && !setttyent())
		return (NULL);
	for (;;) {
		if ((p = fgetln(tf, &len)) == NULL)
			return (NULL);
		if (p[len - 1] != '\n')
			p[len] = '\n';
		while (isspace(*p) && *p != '\n')
			++p;
		if (*p && *p != '\n' && *p != '#')
			break;
	}

	zapchar = 0;
	tty.ty_name = p;
	p = skip(p);
	if (!*(tty.ty_getty = p))
		tty.ty_getty = tty.ty_type = NULL;
	else {
		p = skip(p);
		if (!*(tty.ty_type = p))
			tty.ty_type = NULL;
		else
			p = skip(p);
	}
	tty.ty_status = 0;
	tty.ty_window = NULL;
	tty.ty_auth = NULL;

#define	scmp(e)	!strncmp(p, e, sizeof(e) - 1) && isspace(p[sizeof(e) - 1])
#define	vcmp(e)	!strncmp(p, e, sizeof(e) - 1) && p[sizeof(e) - 1] == '='
	for (; *p; p = skip(p)) {
		if (scmp(_TTYS_OFF))
			tty.ty_status &= ~TTY_ON;
		else if (scmp(_TTYS_ON))
			tty.ty_status |= TTY_ON;
		else if (scmp(_TTYS_DIALOUT))
			tty.ty_status |= TTY_DIALOUT;
		else if (scmp(_TTYS_DIALIN))
			tty.ty_status |= TTY_DIALIN;
		else if (scmp(_TTYS_BIDIR))
			tty.ty_status |= TTY_BIDIR;
		else if (scmp(_TTYS_SECURE))
			tty.ty_status |= TTY_SECURE;
		else if (vcmp(_TTYS_WINDOW))
			tty.ty_window = value(p);
		else if (vcmp(_TTYS_AUTH))
			tty.ty_auth = value(p);
		else
			break;
	}

	if (zapchar == '#' || *p == '#')
		while ((c = *++p) == ' ' || c == '\t')
			;
	tty.ty_comment = p;
	if (*p == '\0')
		tty.ty_comment = NULL;
	if ((p = index(p, '\n')) != NULL)
		*p = '\0';
	return (&tty);
}

#define	QUOTED	1

/*
 * Skip over the current field, removing quotes, and return a pointer to
 * the next field.
 */
static char *
skip(p)
	register char *p;
{
	register char *t;
	register int c, q;

	for (q = 0, t = p; (c = *p) != '\0'; p++) {
		if (c == '"') {
			q ^= QUOTED;	/* obscure, but nice */
			continue;
		}
		if (q == QUOTED && *p == '\\' && *(p+1) == '"')
			p++;
		*t++ = *p;
		if (q == QUOTED)
			continue;
		if (c == '#' || c == '\n') {
			zapchar = c;
			*p = 0;
			break;
		}
		if (c == '\t' || c == ' ') {
			zapchar = c;
			*p++ = 0;
			while ((c = *p) == '\t' || c == ' ')
				p++;
			break;
		}
	}
	*--t = '\0';
	return (p);
}

static char *
value(p)
	register char *p;
{

	return ((p = index(p, '=')) ? ++p : NULL);
}

int
setttyent()
{

	if (tf) {
		(void)rewind(tf);
		return (1);
	} else if ((tf = fopen(_PATH_TTYS, "r")) == NULL)
		return (0);
	fcntl(fileno(tf), F_SETFD, FD_CLOEXEC);
	return (1);
}

int
endttyent()
{
	int rval;

	if (tf) {
		rval = !(fclose(tf) == EOF);
		tf = NULL;
		return (rval);
	}
	return (1);
}
